src/util/requireHook.js

```diff
@@ -1,25 +1,87 @@
-  Module._load = function(file) {
-    var moduleExports = origLoad.apply(Module, arguments);
-    listeners.emit(file, moduleExports);
-    return moduleExports;
+  Module._load = patchedModuleLoad;
+};
+
+function patchedModuleLoad(moduleName) {
+  var i;
+
+  // First attempt to always get the module via the original implementation
+  // as this action may fail. The original function populates the module cache.
+  var moduleExports = origLoad.apply(Module, arguments);
+  var filename = Module._resolveFilename.apply(this, arguments);
+
+  // We are not directly manipulating the global module cache because there might be other tools fiddling with
+  // Module._load. We don't want to break any of them.
+  var cacheEntry = executedHooks[filename] = executedHooks[filename] || {
+    moduleExports: moduleExports,
+
+    // We might have already seen and processed, i.e. manipulated, this require statement. This is something we
+    // are checking using these fields.
+    appliedByModuleNameTransformers: [],
+    byFileNamePatternTransformersApplied: false
   };
+
+  var applicableByModuleNameTransformers = byModuleNameTransformers[moduleName];
+  if (applicableByModuleNameTransformers && cacheEntry.appliedByModuleNameTransformers.indexOf(moduleName) === -1) {
+    for (i = 0; i < applicableByModuleNameTransformers.length; i++) {
+      cacheEntry.moduleExports = applicableByModuleNameTransformers[i](cacheEntry.moduleExports) ||
+        cacheEntry.moduleExports;
+    }
+    cacheEntry.appliedByModuleNameTransformers.push(moduleName);
+  }
+
+  if (!cacheEntry.byFileNamePatternTransformersApplied) {
+    for (i = 0; i < byFileNamePatternTransformers.length; i++) {
+      if (byFileNamePatternTransformers[i].pattern.test(filename)) {
+        cacheEntry.moduleExports = byFileNamePatternTransformers[i].fn(cacheEntry.moduleExports) ||
+          cacheEntry.moduleExports;
+      }
+    }
+    cacheEntry.byFileNamePatternTransformersApplied = true;
+  }
+
+  return cacheEntry.moduleExports;
+}
+
+exports.teardownForTestPurposes = function() {
+  Module._load = origLoad;
+  executedHooks = {};
+  byModuleNameTransformers = {};
 };
 
-exports.teardown = function() {
-  Module._load = origLoad || Module._load;
+exports.onModuleLoad = function on(moduleName, fn) {
+  var eventListeners = byModuleNameTransformers[moduleName] = byModuleNameTransformers[moduleName] || [];
+  eventListeners.push(fn);
 };
 
-exports.on = function on(moduleName, fn) {
-  listeners.once(moduleName, fn);
+exports.onFileLoad = function onFileLoad(pattern, fn) {
+  byFileNamePatternTransformers.push({
+    pattern: pattern,
+    fn: fn
+  });
 };
+
+exports.buildFileNamePattern = function buildFileNamePattern(arr) {
+  return new RegExp(arr.reduce(buildFileNamePatternReducer, ''));
+};
+
+function buildFileNamePatternReducer(agg, pathSegment, i, arr) {
+  if (agg.length > 0) {
+    agg += '\\' + path.sep;
+  }
+  agg += pathSegment;
+  if (i + 1 === arr.length) {
+    agg += '$';
```

Comment:
If we terminate the regex with `$`, should we maybe also start it with `^`? 

Also, the final `$` could simply be added directly in the `buildFileNamePattern` function instead of here in the reducer function, that is, 

`return new RegExp(arr.reduce(buildFileNamePatternReducer, '') + '$');`

This way we don't need to check `i` versus array length...

Reply:
done!

> should we maybe also start it with ^?

This is not possible, because this part is dynamic, i.e. an application could have it's code under `/opt`, `/app`, `C:\`…. The idea here is to always to a (very specific) suffix matching.

Reply:
I changed the express.js test case and example so that this is clearer.